mlfinlab/bet_sizing/ef3m.py

```diff
@@ -0,0 +1,394 @@
+
+            # We only want this to execute once at most, so call a final break if one hasn't been called yet.
+            break
+
+        # Check to see if every value made it through.
+        if len(param_list) < 5:
+            return []
+
+        return param_list
+
+    def single_fit_loop(self, epsilon=10**-5, factor=5, variant=1, max_iter=100_000):
+        """
+        A single scan through the list of mu_2 values, cataloging the successful fittings in a DataFrame.
+
+        :param epsilon: (float) Fitting tolerance
+        :param factor: (float) Lambda factor from equations
+        :param variant: (int) The EF3M variant to execute, options are 1: EF3M using first 4 moments, 2: EF3M using
+        first 5 moments
+        :param max_iter: (int) Maximum number of iterations to perform in the 'fit' method
+        :return: (pd.DataFrame) Fitted parameters and error
+        """
+        # Reset parameters and error for each single_fit_loop.
+        self.parameters = [0 for _ in range(5)]  # Initialize the parameter list.
+        self.error = sum([self.moments[i]**2 for i in range(len(self.moments))])
+
+        std_dev = centered_moment(self.moments, 2)**0.5
+        mu_2 = [float(i) * epsilon * factor * std_dev + self.moments[0] for i in range(1, int(1/epsilon))]
+        err_min = self.error
+
+        d_results = {}
+        for mu_2_i in mu_2:
+            self.fit(mu_2=mu_2_i, epsilon=epsilon, variant=variant, max_iter=max_iter)
+
+            if self.error < err_min:
+                err_min = self.error
+                d_results['mu_1'], d_results['mu_2'], d_results['sigma_1'], d_results['sigma_2'], d_results['p_1'] = \
+                    [[p] for p in self.parameters]
+                d_results['error'] = [err_min]
+
+        return pd.DataFrame.from_dict(d_results)
+
+    def mp_fit(self, epsilon=10**-5, factor=5, n_runs=1, variant=1, max_iter=100_000, num_workers=-1):
+        """
+        Parallelized implementation of the 'single_fit_loop' method. Makes use of dask.delayed to execute multiple
+        calls of 'single_fit_loop' in parallel.
+
+        :param epsilon: (float) Fitting tolerance
+        :param factor: (float) Lambda factor from equations
+        :param n_runs: (int) Number of times to execute 'singleLoop'
+        :param variant: (int) The EF3M variant to execute, options are 1: EF3M using first 4 moments, 2: EF3M using
+         first 5 moments
+        :param max_iter: (int) Maximum number of iterations to perform in the 'fit' method
+        :param num_workers: (int) Number of CPU cores to use for multiprocessing execution. Default is -1 which sets
+         num_workers to all cores.
+        :return: (pd.DataFrame) Fitted parameters and error
+        """
+        # Create a list of delayed objects that return a pd.DataFrame from 'single_fit_loop'.
+        dfs = [delayed(self.single_fit_loop)(epsilon=epsilon, factor=factor, variant=variant, max_iter=max_iter)
```

Comment:
Can we use mlfinlab multiprocessing module instead of Dask? Dask is quite substantial extra dependency, can we avoid that ?

Reply:
Are we trying to keep the number of dependencies to a minimum? I always thought dask was a well-maintained and popular package, and the `delayed` behavior was nice and predictable. I can definitely modify the class to use the multiprocessing module, but I'm just curious as to why you want to avoid dask.

Reply:
As far as I know it is a good convention to minimize the number of dependencies, in this case package installation won't influence versions and cause conflicts in other packages installed in user's environment. Look at pandas and sklearn for example, the number of dependencies is extremely small for such big packages. Furthermore, dask which I personally also find as wonderful package has minimum dependency. If we want to be massively used, minimum dependency is one of the things we should keep in mind, as for me. Secondly, we already have multiprocessing part and it would be nice to use already implemented pieces of code, in this case library code looks more laconic.

Reply:
Makes sense. I'll see what I can do.

Reply:
Done. It was actually simpler to just code using mp.Pool than to change the ef3m functions to be able to make use of mp_pandas_obj. Dask dependency removed.